### ISSUE TITLE
error when trying to futher differentiate using Zygote

### ISSUE THREAD

**Issue by sustc11810424** [JuliaDiff/FiniteDifferences.jl#176]
Summary: it seems that the finite difference can't be further differentiated by Zygote
Example: If I do 
```julia
dif(x) = central_fdm(2, 1)(sin, x)
```
and then call
```julia
dif'(0)
```
I get the following error message:
```julia
setindex!(::StaticArrays.SVector{3, Float64}, value, ::Int) is not defined.

error(::String)@error.jl:33
setindex!(::StaticArrays.SVector{3, Float64}, ::Float64, ::Int64)@indexing.jl:3
#637@array.jl:318[inlined]
#2710#back@adjoint.jl:59[inlined]
Pullback@methods.jl:385[inlined]
(::typeof(∂(_estimate_magnitudes)))(::Tuple{Float64, Nothing})@interface2.jl:0
Pullback@methods.jl:365[inlined]
(::typeof(∂(estimate_step)))(::Tuple{Float64, Nothing})@interface2.jl:0
Pullback@methods.jl:193[inlined]
(::typeof(∂(λ)))(::Float64)@interface2.jl:0
Pullback@Other: 1[inlined]
(::typeof(∂(dif)))(::Float64)@interface2.jl:0
(::Zygote.var"#41#42"{typeof(∂(dif))})(::Float64)@interface.jl:41
gradient(::Function, ::Int64)@interface.jl:59
(::Zygote.var"#43#44"{typeof(Main.workspace50.dif)})(::Int64)@interface.jl:62
top-level scope@Local: 1[inlined]
```

The actual use case is that I'm trying to train a NN where I need to compute the gradient of its parameters w.r.t a loss function that involves finite differences given by ```central_fdm```. Is this possible?

**Comment by simeonschaub** [comment]
Can you just use it the other way, i.e. FiniteDifferences over Zygote instead of Zygote over FiniteDifferences? You could actually define a custom adjoint for `FiniteDifferenceMethod` which does this automatically for cases like this.

**Comment by wesselb** [comment]
Step size adaptation appears to be not working well with Zygote. Unfortunately, I'm not familiar with the inner workings of Zygote to see what precisely breaks down. Perhaps @willtebbutt or @oxinabox could provide some insight. From the error message, if appears that a pullback attempts to modify a `StaticArray` in-place, which won't work.

Once you turn of step size adaptation, thing seem to work:

```julia
julia> fdm = central_fdm(5, 1; adapt=0);

julia> cos_(x) = fdm(sin, x)
cos_ (generic function with 1 method)

julia> cos_'(1)
-0.8414709848079269

julia> -sin(1)
-0.8414709848078965
```

I would be careful with AD-ing through finite difference estimates, though. I second @simeonschaub's suggestion of taking finite differences of gradients computed by AD.
